zing_stats.py

```diff
@@ -164,140 +180,212 @@ def main():
-
-        # and add an all group for all projects
-        group_map['All'] = proj_map.keys()
-        log.debug('group_map: %s', group_map)
-
-        for group in group_map:
-            projects = group_map[group]
-            groups = group_map.keys()
-            html = generate_html(args, df, num_changes, start_dt, finish_dt,
-                                 projects, all_projects, group, groups)
-            dir_path = os.path.join(args.output_dir, file_prefix)
-            if group == 'All':
-                file_name = 'index.html'
-            else:
-                file_name = '%s.html' % group.lower()
-            file_path = os.path.join(dir_path, file_name)
-            if not os.path.exists(dir_path):
-                os.makedirs(dir_path)
-            with open(file_path, 'w') as f:
-                f.write(html)
-            log.info('Wrote %s for %s group', file_path, group)
-    else:
+    all_projects = teams_map['All']
+    for team in sorted(teams_map):
+        team_projects = teams_map[team]
+        teams = teams_map.keys()
+        html = generate_html(args, df, num_changes, start_dt, finish_dt,
+                             team_projects, all_projects, team, teams)
         dir_path = os.path.join(args.output_dir, file_prefix)
-        file_name = 'index.html'
+        if team == 'All':
+            file_name = 'index.html'
+        else:
+            file_name = '%s.html' % re.sub(r'\W+', '_', team.lower())
         file_path = os.path.join(dir_path, file_name)
         if not os.path.exists(dir_path):
             os.makedirs(dir_path)
-        html = generate_html(args, df, num_changes, start_dt, finish_dt,
-                             df.keys(), all_projects, 'All')
         with open(file_path, 'w') as f:
             f.write(html)
-        log.info('Wrote %s', file_path)
+        log.info('Wrote %s for team: "%s"', file_path, team)
+
+
+def gather_github_prs(args, oldest_timestamp, projects):
+    projects_github = [x['name'] for x in projects.get('github', dict())]
+
+    if len(projects_github) < 1:
+        return 0, dict()
+
+    log.info('Gathering PRs from %s for %s', args.github_url,
+             ', '.join(sorted(projects_github)))
+    total_prs = 0
+    prs = dict()
+    for project in projects_github:
+        prs[project] = dict()
+
+        project_finished = False
+        next_page = True
+
+        payload = {'state': 'all', 'sort': 'updated', 'direction': 'dsc'}
+        query = ('%s/api/v3/repos/%s/pulls'
+                 % (args.github_url, project))
+        while next_page:
+            if args.github_user and args.github_token:
+                response = requests.get(
+                    query, params=payload,
+                    verify=args.verify_https_requests,
+                    auth=(args.github_user, args.github_token))
+            else:
+                response = requests.get(query, params=payload,
+                                        verify=args.verify_https_requests)
```

Comment:
Consider creating a session object outside the loop, should help increase performance, and if you then want to move this requests code to a separate function, it should be possible to run multiple queries in parallel with some simple threading code (jenkins-job-builder has some code for doing something similar) or a library to dispatch and return the result.

Reply:
Very good point - I've a patch for this under test which I intend to land as a separate PR after this. On a test run against a mix of 300 gerrit and github projects, using requests session reduces the runtime from 11 minutes to 6 minutes - so very much worth doing.